python/parquet_tools/common/schema.py

```diff
@@ -0,0 +1,700 @@
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+This module encapsulates all the code to process schema files and provide
+support for schemas that allow null values for non-floating point values.
+
+This module augments schemas that allow null values for data types which
+pandas does not support to be null. Specifically, for each such column it
+adds a  column with a name of '<cname>_ISNULL', where <cname> is the name
+of the column that may contain an unsupported null value. The associated
+column will contain true if the original column contains a null value or
+false, otherwise. Such a column will contain a well defined value or a
+value specified by the command issuer. While this module sets up the
+mechanics, the realization of this process is handled by the caller.
+"""
+
+import csv
+import os
+import numpy as np
+
+from ..common.typeinfo import TypeInfo
+from ..common.errhandlers import ErrInfo, FatalError, ParmError
+
+
+# *****************************************************************************
+# *                             S c h e m a . p y                             *
+# *****************************************************************************
+
+class Schema:
+
+    # Public column information
+    #
+    colNames = None   # List of column names (None if not a defined schema)
+    colNVChk = []     # The columns that need to be checked for a null value
+    colTypes = {}     # Dictionary of column name to pandas type
+    colOrigs = 0      # Original number of columns in schema
+
+    # Private members
+    #
+    _cNameMaw = 0     # The length of the longest column name in original
+    _cNameMax = 0     # The length of the longest column name in augmented
+    _colFmtfp = ''    # Current format in effect when formating floats
+    _colFPFmt = []    # The columns that need to be formatted as fixed point
+    _colFP32 = []     # List of column names with type float.
+    _colFP64 = []     # List of column names with type double.
+    _colNotNull = []  # List of column names that cannot be null
+    _colSpecs = []    # List of [column name, SQL type, pandas type]
```

Comment:
I would prefer a class or named tuple for a list of coupled types.

Reply:
Well, these are not pair-able and are built separately. So, while they, as a whole, describe the schema they are not uniform. Making it an n-tuple would be a big headache and difficult to deal with.